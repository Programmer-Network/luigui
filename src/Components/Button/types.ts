import { ReactNode } from "react";

export interface IButtonProps {
  onClick?: (e: React.MouseEvent) => void;
  children?: ReactNode;
  className?: string;
  disabled?: boolean;
  type?: "button" | "submit" | "reset";
  isLoading?: boolean;
  outlined?: boolean;
}
