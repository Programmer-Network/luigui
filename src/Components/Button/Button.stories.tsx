import type { Meta, StoryObj } from "@storybook/react";

import Button from ".";

const meta = {
  title: "Core/Button",
  component: Button,
  parameters: {
    layout: "centered"
  },
  tags: ["autodocs"],
  argTypes: {}
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Filled: Story = {
  args: {
    children: "Button"
  }
};

export const FilledDisabled: Story = {
  args: {
    children: "Button",
    disabled: true
  }
};

export const FilledSpining: Story = {
  args: {
    children: "Button",
    isLoading: true
  }
};

export const Outlined: Story = {
  args: {
    children: "Button",
    outlined: true
  }
};

export const OutlinedDisabled: Story = {
  args: {
    children: "Button",
    outlined: true,
    disabled: true
  }
};

export const OutlinedSpinning: Story = {
  args: {
    children: "Button",
    outlined: true,
    isLoading: true
  }
};

export const WithChangedCustomCSSProperty: Story = {
  args: {
    children: "Button",
    outlined: true,
    isLoading: true,
    className: "yail-button--error"
  }
};
