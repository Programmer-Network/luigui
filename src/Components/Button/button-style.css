.button--red {
  --color-primary: 220 38 38;
}

.button--gray {
  --color-primary: 107 114 128;
}

.button--green {
  --color-primary: 22 163 74;
}
