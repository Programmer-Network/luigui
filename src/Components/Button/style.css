.yail-button--error {
  --color-primary: #dc2626;
}
