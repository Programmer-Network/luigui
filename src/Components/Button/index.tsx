import classNames from "classnames";

import Spinner from "Components/Spinner";

import "./style.css";
import { IButtonProps } from "./types";

const Button: React.FC<IButtonProps> = ({
  onClick,
  children,
  className,
  disabled = false,
  type = "button",
  isLoading = false,
  outlined = false
}) => {
  return (
    <button
      disabled={disabled}
      onClick={onClick}
      type={type}
      className={classNames(
        "select-none border-2 border-primary px-3 py-2 text-sm font-semibold uppercase tracking-tight shadow-lg",
        className,
        {
          "cursor-not-allowed opacity-70": disabled,
          "text-primary": outlined && !disabled,
          "text-primary-background-color": !outlined || disabled,
          "!text-bg-primary-background-color !bg-primary":
            !outlined && disabled,
          "hover:bg-transparent": !outlined && isLoading,
          "bg-primary hover:bg-transparent hover:text-primary":
            !outlined && !disabled,
          "hover:bg-primary hover:text-primary-background-color":
            outlined && !disabled,
          "!bg-transparent !text-primary": outlined && disabled
        }
      )}
    >
      <div className='relative flex items-center justify-center'>
        <span
          className={classNames({
            invisible: isLoading
          })}
        >
          {children}
        </span>
        <span className='absolute'>
          {isLoading && <Spinner className={classNames("h-8 w-8")} />}
        </span>
      </div>
    </button>
  );
};

Button.defaultProps = {
  onClick: () => null,
  children: null,
  className: "",
  disabled: false,
  outlined: false,
  isLoading: false
};

export default Button;
