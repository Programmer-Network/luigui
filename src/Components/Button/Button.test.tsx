import { fireEvent, render, screen } from "@testing-library/react";
import { vi } from "vitest";

import Button from ".";

describe("Button component", () => {
  test("renders correctly - snapshot test", () => {
    const { asFragment } = render(<Button>Snapshot Button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders button text", () => {
    render(<Button>Click me</Button>);
    expect(screen.getByText("Click me")).toBeInTheDocument();
  });

  test("is disabled when disabled prop is true", () => {
    render(<Button disabled>Click me</Button>);
    const button = screen.getByRole("button");
    expect(button).toBeDisabled();
  });

  test("shows spinner when isLoading is true", () => {
    render(<Button isLoading={true}>Click me</Button>);
    expect(screen.getByTestId("icon-spinner")).toBeInTheDocument();
  });

  test("has outlined styles when variant is outlined", () => {
    render(<Button variant='outlined'>Click me</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("yl-border-2");
    expect(button).toHaveClass("yl-text-primary");
    expect(button).toHaveClass("yl-border-primary");
  });

  test("has text styles when variant is text", () => {
    render(<Button variant='text'>Click me</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("yl-border-transparent");
    expect(button).toHaveClass("yl-text-primary");
    expect(button).toHaveClass("hover:yl-bg-primary/10");
  });

  test("handles click events", () => {
    const handleClick = vi.fn();
    render(<Button onClick={handleClick}>Click me</Button>);
    fireEvent.click(screen.getByText("Click me"));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test("has correct styles when disabled and filled", () => {
    render(<Button disabled>Disabled Button</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("yl-text-primary-background-color");
    expect(button).toHaveClass("yl-bg-primary");
  });

  test("has correct styles when loading and filled", () => {
    render(<Button isLoading>Button</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("hover:yl-bg-transparent");
  });

  test("has correct styles when not disabled, not loading, and filled", () => {
    render(<Button>Button</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("yl-bg-primary");
    expect(button).toHaveClass("hover:yl-bg-transparent");
    expect(button).toHaveClass("hover:yl-text-primary");
  });

  test("has correct styles when disabled and outlined", () => {
    render(
      <Button disabled variant='outlined'>
        Button
      </Button>
    );
    const button = screen.getByRole("button");
    expect(button).toHaveClass("yl-text-primary");
  });

  test("has correct styles when disabled and text variant", () => {
    render(
      <Button disabled variant='text'>
        Button
      </Button>
    );
    const button = screen.getByRole("button");
    expect(button).toHaveClass("yl-text-primary");
    expect(button).toHaveClass("yl-border-transparent");
    expect(button).toHaveClass("yl-opacity-70");
  });

  test("has correct styles when loading and text variant", () => {
    render(
      <Button isLoading variant='text'>
        Button
      </Button>
    );
    const button = screen.getByRole("button");
    expect(button).toHaveClass("yl-border-transparent");
    expect(screen.getByTestId("icon-spinner")).toBeInTheDocument();
  });

  test("has correct hover styles for text variant", () => {
    render(<Button variant='text'>Button</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("hover:yl-bg-primary/10");
    expect(button).toHaveClass("yl-border-transparent");
    expect(button).toHaveClass("yl-text-primary");
  });
});
