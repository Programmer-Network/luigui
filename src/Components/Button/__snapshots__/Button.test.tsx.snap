// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button component > renders correctly - snapshot test 1`] = `
<DocumentFragment>
  <button
    class="select-none border-2 border-primary px-3 py-2 text-sm font-semibold uppercase tracking-tight shadow-lg text-primary-background-color bg-primary hover:bg-transparent hover:text-primary"
    type="button"
  >
    <div
      class="relative flex items-center justify-center"
    >
      <span
        class=""
      >
        Snapshot Button
      </span>
      <span
        class="absolute"
      />
    </div>
  </button>
</DocumentFragment>
`;
